records/generator.go

```diff
@@ -57,8 +67,8 @@ func (rg *RecordGenerator) ParseState(c Config, masters ...string) error {
 
 // Tries each master and looks for the leader
 // if no leader responds it errors
-func (rg *RecordGenerator) findMaster(masters ...string) (state.State, error) {
-	var sj state.State
+func (rg *RecordGenerator) findMaster(masters ...string) (*state.State, error) {
```

Comment:
It really depends on the API. For example, the io.Reader api tells you to
check for data, regardless of whether an error is returned.

We don't need complicated API semantics here because we don't have
different types of errors to handle in different ways. If err != nil then I
think we shouldn't trust the returned state -- end of story.

I don't hate the change to *state.State -- I just don't think it's
necessary.


On Thu, Nov 12, 2015 at 12:05 PM, Sargun Dhillon <notifications@github.com>
wrote:

> In records/generator.go
> <https://github.com/mesosphere/mesos-dns/pull/341#discussion_r44683946>:
>
> > @@ -57,8 +67,8 @@ func (rg *RecordGenerator) ParseState(c Config, masters ...string) error {
> >
> >  // Tries each master and looks for the leader
> >  // if no leader responds it errors
> > -func (rg *RecordGenerator) findMaster(masters ...string) (state.State, error) {
> > -	var sj state.State
> > +func (rg *RecordGenerator) findMaster(masters ...string) (*state.State, error) {
>
> I think my reasoning here was state == nil iff err != nil. If someone is
> checking the err condition, and bailing out, there should be no need for
> them to ever check whether state == nil. I see the benefit as if the value
> is somehow dereferenced, because error checking went awry, it'll crash.
> From my understanding, mesos-dns is typically run under a supervisor, which
> will restart it. I see this as a better alternative than loading in an
> incorrect state. I see this as a variant of crash-only computing.
>
> If you look at the behaviour in the Go standard library, it appears to be
> the same: https://play.golang.org/p/aUUoHoNXnM.
>
> Although, I've noticed in our project, we've followed the semantic of
> preferring to load the zero value. If we'd like to continue that, I don't
> think it's detrimental.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mesosphere/mesos-dns/pull/341/files#r44683946>.
>